sideload.go

```diff
@@ -0,0 +1,420 @@
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightninglabs/neutrino/chaindataloader"
+	"github.com/lightninglabs/neutrino/chainsync"
+	"github.com/lightninglabs/neutrino/headerfs"
+	"github.com/lightninglabs/neutrino/headerlist"
+)
+
+const (
+	sideloadDataChunkSize int32 = 2000
+)
+
+// sideloader a struct that contains all dependencies required to fetch and
+// process sideloaded headers.
+type sideloader struct {
+	// reader is the sideload source.
+	blkHdrReader chaindataloader.Reader
+
+	cfHdrReader chaindataloader.Reader
+	// blkHdrProcessor contains dependies required to verify and store block
+	// headers.
+	blkHdrProcessor *blkHdrProcessor
+
+	// blkHdrSkipVerify indicates if processing the headers require verification.
+	blkHdrSkipVerify bool
+
+	fHdrSkipVerify bool
+
+	// fHdrNextCheckpointIdx indicates the index of the next filter header
+	// checkpoint height relative to the current filter tip.
+	fHdrNextCheckpointIdx int
+
+	// fHdrStore is the store for filter headers.
+	fHdrStore *headerfs.FilterHeaderStore
+
+	// filterCheckpoints is the map of hardcoded filter checkpoints to their
+	//hash.
+	filterCheckpoints map[uint32]*chainhash.Hash
+
+	// filterCheckptHeights is the slice of hardcoded filter checkpoint heights.
+	filterCheckptHeights []uint32
+
+	// blkHdrCurHeight is the height of the block header's store tip.
+	blkHdrCurHeight int32
+
+	// fHdrCurHeight is the height of the filter header's store tip.
+	fHdrCurHeight int32
+
+	// fHdrCurHash is the filter hash of the filter header's store tip.
+	fHdrCurHash *chainhash.Hash
+
+	// fHdrNextCheckptHeight is the height of the next filter header next
+	// checkpoint's height relative to the current filter header tip.
+	fHdrNextCheckptHeight int32
+
+	// fHdrLastHeight is the last height to which we can sideload the filter
+	// headers to. This restriction occurs because the filter header tip
+	// cannot surpass the block header tip as it depends on it.
+	fHdrLastHeight uint32
+}
+
+// processSideloadedCfHeader verifies and stores the passed cfheaders.
+// It returns the current height and next checkpoint after this process to be
+// used by the called. It returns an error where necessary.
+func (h *sideloader) processSideloadedCfHeader(
+	headers []*chainhash.Hash) (*chaindataloader.ProcessHdrResp, error) {
+
+	msgCfHeader := &wire.MsgCFHeaders{
+		FilterType:       fType,
+		StopHash:         *headers[len(headers)-1],
+		PrevFilterHeader: *h.fHdrCurHash,
+		FilterHashes:     headers,
+	}
+
+	// We only verify, if this is set to false.
+	if !h.fHdrSkipVerify {
+
+		if !verifyCheckpoint(h.fHdrCurHash, headers[len(headers)-1],
+			msgCfHeader) {
+
+			log.Warnf("Filter headers failed verification at height  - %v",
+				h.fHdrCurHeight)
+			return nil, nil
+		}
+		h.fHdrNextCheckpointIdx = h.fHdrNextCheckpointIdx + 1
+		h.fHdrNextCheckptHeight = int32(h.filterCheckptHeights[h.
+			fHdrNextCheckpointIdx])
+
+	}
+
+	_, _, _, err := writeCfHeaders(msgCfHeader, h.fHdrStore,
```

Comment:
I think we just need to go one step fruther to move logic like this into the generic struct.